File: backend/app/schemas/workspace.py
"""Workspace API schemas."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class WorkspaceSessionCreateRequest(BaseModel):
    purpose: str = "skill_create"
    seed_prompt: str = ""
    skill_id: str = ""
    reuse_existing: bool = True


class WorkspaceSessionCreateResponse(BaseModel):
    workspace_id: str
    session_id: str
    status: str
    display_root: str
    real_vfs_root: str
    skill_id: str = ""


class WorkspaceSessionStatusResponse(BaseModel):
    workspace_id: str
    session_id: str
    status: str
    display_root: str
    real_vfs_root: str
    created_at: str
    skill_id: str = ""


class WorkspaceRunSessionCreateResponse(BaseModel):
    workspace_id: str
    run_session_id: str
    skill_id: str = ""


class WorkspaceTreeNode(BaseModel):
    name: str
    path: str
    node_type: str
    children: list["WorkspaceTreeNode"] = Field(default_factory=list)


class WorkspaceTreeResponse(BaseModel):
    workspace_id: str
    root: str
    nodes: list[WorkspaceTreeNode]


class WorkspaceFileResponse(BaseModel):
    path: str
    content: str
    readonly: bool = False
    content_type: str = "text/plain"
    size_bytes: int = 0
    etag: str = ""
    truncated: bool = False
    is_binary: bool = False


class WorkspaceFileUpdateRequest(BaseModel):
    path: str
    content: str
    if_match_etag: str = ""


class WorkspaceFileUpdateResponse(BaseModel):
    ok: bool
    path: str
    bytes: int
    etag: str = ""


class WorkspaceCreatorResultResponse(BaseModel):
    found: bool
    result: dict[str, Any] | None = None


WorkspaceTreeNode.model_rebuild()

File: frontend/src/types/workspace.ts
export type WorkspaceStatus = "initializing" | "ready" | "failed";

export interface WorkspaceSession {
  workspace_id: string;
  session_id: string;
  status: WorkspaceStatus;
  display_root: string;
  real_vfs_root: string;
  skill_id?: string;
}

export interface WorkspaceRunSession {
  workspace_id: string;
  run_session_id: string;
  skill_id?: string;
}

export interface WorkspaceTreeNode {
  name: string;
  path: string;
  node_type: "file" | "dir";
  children: WorkspaceTreeNode[];
}

export interface WorkspaceTreeResponse {
  workspace_id: string;
  root: string;
  nodes: WorkspaceTreeNode[];
}

export interface WorkspaceFilePayload {
  path: string;
  content: string;
  readonly: boolean;
  content_type: string;
  size_bytes: number;
  etag: string;
  truncated: boolean;
  is_binary: boolean;
}

export interface WorkspaceFileUpdatePayload {
  ok: boolean;
  path: string;
  bytes: number;
  etag: string;
}

export interface WorkspaceCreatorResultPayload {
  found: boolean;
  result: Record<string, unknown> | null;
}

export interface WorkspaceTerminalLineEvent {
  stream: "stdout" | "stderr";
  line: string;
  tool_id?: string;
  run_id?: string;
  ts?: string;
}

export interface WorkspaceFsChangedEvent {
  op: "created" | "updated" | "deleted";
  path: string;
  ts?: string;
}
